test/src/test/java/org/corfudb/integration/AddNodeIT.java

```diff
@@ -0,0 +1,131 @@
+
+        // Start node one and populate it with data
+        new CorfuServerRunner()
+                .setHost(host)
+                .setPort(n1Port)
+                .setSingle(true)
+                .runServer();
+
+        CorfuRuntime n1Rt = new CorfuRuntime(getConnectionString(n1Port)).connect();
+
+        SMRMap<String, String> map = n1Rt.getObjectsView()
+                .build()
+                .setType(SMRMap.class)
+                .setStreamName(streamName)
+                .open();
+
+        final int numEntris = 10_000;
+        for (int x = 0; x < numEntris; x++) {
+            map.put(String.valueOf(x), String.valueOf(x));
+        }
+
+        // Add a second node
+        final int n2Port = 9001;
+        new CorfuServerRunner()
+                .setHost(host)
+                .setPort(n2Port)
+                .runServer();
+
+        ManagementClient mgmt = n1Rt.getRouter(getConnectionString(n1Port))
+                .getClient(ManagementClient.class);
+
+        AddNodeResponse resp = mgmt.addNodeRequest(getConnectionString(n2Port));
+
+        assertThat(resp.getWorkflowId()).isNotNull();
+
+        waitForWorkflow(resp.getWorkflowId(), n1Rt, n1Port);
+
+        n1Rt.invalidateLayout();
+        final int clusterSizeN2 = 2;
+        assertThat(n1Rt.getLayoutView().getLayout().getAllServers().size()).isEqualTo(clusterSizeN2);
+
+        MultiCheckpointWriter mcw = new MultiCheckpointWriter();
+        mcw.addMap(map);
+
+        long prefix = mcw.appendCheckpoints(n1Rt, "Maithem");
+
+        n1Rt.getAddressSpaceView().prefixTrim(prefix - 1);
+
+        n1Rt.getAddressSpaceView().invalidateClientCache();
+        n1Rt.getAddressSpaceView().invalidateServerCaches();
+        n1Rt.getAddressSpaceView().gc();
+
+        // Add a 3rd node after compaction
+
+        final int n3Port = 9002;
+        new CorfuServerRunner()
+                .setHost(host)
+                .setPort(n3Port)
+                .runServer();
+
+        AddNodeResponse resp2 = mgmt.addNodeRequest(getConnectionString(n3Port));
+        assertThat(resp2.getWorkflowId()).isNotNull();
+
+        waitForWorkflow(resp2.getWorkflowId(), n1Rt, n1Port);
+
+        // Verify that the third node has been added and data can be read back
+        n1Rt.invalidateLayout();
+        final int clusterSizeN3 = 3;
+        assertThat(n1Rt.getLayoutView().getLayout().getAllServers().size()).isEqualTo(clusterSizeN3);
+        for (int x = 0; x < numEntris; x++) {
+            String v = map.get(String.valueOf(x));
+            assertThat(v).isEqualTo(String.valueOf(x));
+        }
+    }
+
+    void waitForWorkflow(UUID id, CorfuRuntime rt, int port) throws Exception {
+        ManagementClient mgmt = rt.getRouter(getConnectionString(port))
+                .getClient(ManagementClient.class);
+        for (int x = 0; x < maxTries; x++) {
+            try {
+                if (mgmt.queryRequest(id).isActive()) {
+                    Thread.sleep(sleepTime);
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Remove this use of "Thread.sleep()". [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=squid%3AS2925)